### PR TITLE
Added heapster to set of addons

### PR DESCRIPTION
**Pull request by aaron-prindle**
Ref: #743

### CODE REVIEW

**Comment by aaron-prindle** [comment]
Right now, grafana can be accessed via:
`m service -n kube-system monitoring-grafana`
should there be a convenience method make like the dashboard?
`minikube heapster` ?

**Comment by r2d4** [comment]
Do we have a way to disable this by default?  Additionally, do we want it on or off by default?

Does hpa (http://kubernetes.io/docs/user-guide/horizontal-pod-autoscaling/) work in minikube now when you have heapster on?

**Comment by aaron-prindle** [comment]
Yes, it can be disabled by default in addons.go, it currently is enabled though (arbitrarily).  I will test hpa and report back.

File: deploy/addons/heapster-rc.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    k8s-app: heapster
+  template:
+    metadata:
+      labels:
+        kubernetes.io/cluster-service: 'true'
+        k8s-app: heapster
+    spec:
+      containers:
+      - name: heapster
+        image: kubernetes/heapster:canary
```
Comment (r2d4) [inline deploy/addons/heapster-rc.yaml]: We should tie this to a release instead of canary
Maybe 0.19?

https://hub.docker.com/r/kubernetes/heapster/tags/

Comment (aaron-prindle) [inline deploy/addons/heapster-rc.yaml]: Sure, I was just following what official heapster repo did.  I can change to 0.19

**Comment by aaron-prindle** [comment]
To login to grafana, you have to maneuver through the UI and there is also a username/password prompt (admin/admin).  Should there be display stating to this when you use the convenience command?  If so how should this work with the --url command (should this help text always be displayed)?

**Comment by r2d4** [comment]
For setting up and logging in, I think we could probably just add a section to the readme linking to the heapster docs.

Also, the `addon <name> --url` command assumes that each addon has a single corresponding service named after the key we've given it in our addon map.  I don't think its unreasonable that it does (all of our current addons and ingress does) -- but it doesn't have to.

**Comment by r2d4** [comment]
Thinking a little more, adding a label to the service is a pretty clean solution.  Then we can just lookup all services in kube-namespace with label `heapster` and print out their urls.  

That would support the 0, 1, 'many' cases.

File: deploy/addons/heapster-rc.yaml (hunk 1 above)
Comment (r2d4) [inline deploy/addons/heapster-rc.yaml]: So it turns out they have been releasing to GCR but pushing latest to dockerhub through their makefiles.  I opened up an issue here https://github.com/kubernetes/heapster/issues/1384

But we should probably pin to their last release
gcr.io/google_containers/heapster:v1.2.0

File: deploy/addons/dashboard-svc.yaml [hunk 2]
```diff
@@ -21,6 +21,8 @@ metadata:
   labels:
     app: kubernetes-dashboard
     kubernetes.io/cluster-service: "true"
+    minikube/addons: dashboard
```
Comment (dlorenc) [inline deploy/addons/dashboard-svc.yaml]: could you put this under kubernetes.io/minikube-addons and kubernetes.io/minikube-addons-endpoint

**Comment by aaron-prindle** [comment]
hpa has been tested and is working:
hpa

**Comment by aaron-prindle** [comment]
PTAL, all points have been addressed.  To test heapster addon, you will have to run:
minikube addons enable heapster
minikube addons open heapster

File: cmd/minikube/cmd/config/open.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			fmt.Fprintln(os.Stderr, "usage: minikube addons open ADDON_NAME")
+			os.Exit(1)
+		}
+		addonName := args[0]
+		api := libmachine.NewClient(constants.Minipath, constants.MakeMiniPath("certs"))
+		defer api.Close()
+
+		cluster.EnsureMinikubeRunningOrExit(api, 1)
+		if _, ok := assets.Addons[addonName]; !ok { // validate addon input
+			fmt.Fprintln(os.Stderr, fmt.Sprintf("addon '%s' is not a valid addon packaged with minikube", addonName))
```
Comment (dlorenc) [inline cmd/minikube/cmd/config/open.go]: Is there a way for the user to list the valid add-ons? Could you add that to this error?

File: cmd/minikube/cmd/config/open.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		if len(args) != 1 {
+			fmt.Fprintln(os.Stderr, "usage: minikube addons open ADDON_NAME")
+			os.Exit(1)
+		}
+		addonName := args[0]
+		api := libmachine.NewClient(constants.Minipath, constants.MakeMiniPath("certs"))
+		defer api.Close()
+
+		cluster.EnsureMinikubeRunningOrExit(api, 1)
+		if _, ok := assets.Addons[addonName]; !ok { // validate addon input
```
Comment (dlorenc) [inline cmd/minikube/cmd/config/open.go]: Do we check that the addon is enabled?

File: cmd/minikube/cmd/config/open.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	addonsURLTemplate *template.Template
+)
+
+const defaultAddonsFormatTemplate = "http://{{.IP}}:{{.Port}}"
+
+var addonsOpenCmd = &cobra.Command{
+	Use:   "open ADDON_NAME",
+	Short: "Opens the addon w/ADDON_NAME within minikube (example: minikube addons open dashboard). For a list of available addons use: minikube addons list ",
+	Long:  "Opens the addon w/ADDON_NAME within minikube (example: minikube addons open dashboard). For a list of available addons use: minikube addons list ",
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
```
Comment (dlorenc) [inline cmd/minikube/cmd/config/open.go]: Should this just be PreRun instead of persistent?

File: pkg/minikube/cluster/cluster.go [hunk 6]
```diff
@@ -683,3 +685,86 @@ func GetServiceURLs(api libmachine.API, namespace string, t *template.Template)
...
+		return &util.RetriableError{Err: err}
+	}
+	endpoint, err := endpoints.Get(service)
+	if err != nil {
+		return &util.RetriableError{Err: err}
+	}
+	return checkEndpointReady(endpoint)
+}
+
+const notReadyMsg = "Waiting, endpoint for service is not ready yet...\n"
```
Comment (dlorenc) [inline pkg/minikube/cluster/cluster.go]: Move this to the top or into a function?

File: cmd/minikube/cmd/config/open.go (hunk 5 above)
Comment (aaron-prindle) [inline cmd/minikube/cmd/config/open.go]: You are correct, I believe it should be PreRun. I will change it in service.go cmd as well .

File: cmd/minikube/cmd/config/open.go (hunk 4 above)
Comment (aaron-prindle) [inline cmd/minikube/cmd/config/open.go]: No, I will add this

File: cmd/minikube/cmd/config/open.go (hunk 3 above)
Comment (aaron-prindle) [inline cmd/minikube/cmd/config/open.go]: I will add a note to run `minikube addons list`

**Comment by aaron-prindle** [comment]
@minikube-bot retest this please

**Pull request merged**